### DISCUSSION TITLE
Multiple Windows and UI Threads/Dispatchers?

### DISCUSSION THREAD

**Discussion by legistek** [AvaloniaUI/Avalonia#11876] [Q&A]
Is it possible without modifying the Avalonia internal source code to have multiple top-level windows, each on their **_own_** UI thread and `Dispatcher`?

If not, does anyone have a general sense of what would be required to modify Avalonia to support this? I see the static `Dispatcher.UIThread` for example might have to be removed and rather each visual would need to have its own `Dispatcher` reference. Any other gotchas?

And for anyone wondering why I would want this, please just take my word for it that I do. :)

Thanks!

**Comment by robloo** [comment]
It's not possible by my understanding. Certain platforms only support single thread for UI as well so this is unlikely to ever change.

I wonder why you need this though? App UI work should be fairly lightweight with background threads or tasks doing most things. Then working with multiple UI threads can get tricky when event order is important. Seems better never to do this.

Finally, there is nothing stopping two instances of the app from running -- although sharing data is harder.

**Reply by legistek** [reply]
>  I wonder why you need this though? 

Commander Lock:
Dammit, Morpheus. Not everyone believes what you believe.

Morpheus:
My beliefs do not require them to.

**Reply by robloo** [reply]
Do as you will, convention and standards are there for a reason but there is always a special case (as unlikely as that is). That said, there is also always other ways to do things within existing constraints.

**Reply by maxkatz6** [reply]
macOS doesn't support multiple UI threads. That's the biggest reason.

**Reply by legistek** [reply]
Yeah the other way is to fork Avalonia and just implement it.

**Reply by legistek** [reply]
Understood, thanks @maxkatz6.

**Comment by MarkBrainfire** [comment]
Of course, we need multithreading / multitasking so that computationally intense background tasks don't interfere with the UI thread.
I'm glad to see that this has been implemented for desktop (Windows) and Android.  However, this hasn't yet been implemented on WASM.  Are there plans for this?

Mark

**Reply by timunie** [reply]
See #12785

**Reply by maxkatz6** [reply]
> we need multithreading / multitasking so that computationally intense background tasks don't interfere with the UI thread.

It's obviously available for all platforms except browser. This topic question is about multiple UI threads, whichs is not possible on some platforms, like macos or ios.

**Reply by MarkBrainfire** [reply]
Got it.

**Comment by dxh-sc** [comment]
WPF has multiple UI threads, so i'm looking forward to this feature in Avalonia.

**Reply by jp2masa** [reply]
This is almost the same as saying "WPF only supports Windows, so I'm looking forward to this feature in Avalonia".

As explained by other comments, at least macOS and iOS don't support this, so it doesn't bring much value to Avalonia. Actually, if these platforms can live without it, you probably can, too. I know IPC is not easy, but thread synchronization isn't much easier either.
